airflow/utils/helpers.py

```diff
@@ -183,87 +183,65 @@ def f(t):
+                    attempting to SIGKILL
+    :type timeout: int
     """
     try:
         root_process = psutil.Process(pid)
     except psutil.NoSuchProcess:
         logger.warning("PID: {} does not exist".format(pid))
         return
 
-    # Check child processes to reduce cases where a child process died but
-    # the PID got reused.
-    descendant_processes = [x for x in root_process.children(recursive=True)
-                            if x.is_running()]
+    # Check child processes to reduce cases where a child process died but the PID
+    # got reused
+    running_descendants = \
+        [x for x in root_process.children(recursive=True) if x.is_running()]
 
-    if len(descendant_processes) != 0:
+    if len(running_descendants) != 0:
         logger.info("Terminating descendant processes of {} PID: {}"
-                    .format(root_process.cmdline(),
-                            root_process.pid))
-        temp_processes = descendant_processes[:]
-        for descendant in temp_processes:
-            logger.info("Terminating descendant process {} PID: {}"
-                        .format(descendant.cmdline(), descendant.pid))
-            if not kill_using_shell(logger, descendant.pid, signal.SIGTERM):
-                descendant_processes.remove(descendant)
-
-        logger.info("Waiting up to {}s for processes to exit..."
-                    .format(timeout))
-        try:
-            psutil.wait_procs(descendant_processes, timeout)
-            logger.info("Done waiting")
-        except psutil.TimeoutExpired:
-            logger.warning("Ran out of time while waiting for "
-                           "processes to exit")
-        # Then SIGKILL
-        descendant_processes = [x for x in root_process.children(recursive=True)
-                                if x.is_running()]
-
-        if len(descendant_processes) > 0:
-            temp_processes = descendant_processes[:]
-            for descendant in temp_processes:
-                logger.info("Killing descendant process {} PID: {}"
-                            .format(descendant.cmdline(), descendant.pid))
-                if not kill_using_shell(logger, descendant.pid, signal.SIGKILL):
-                    descendant_processes.remove(descendant)
-                else:
-                    descendant.wait()
-            logger.info("Killed all descendant processes of {} PID: {}"
-                        .format(root_process.cmdline(),
-                                root_process.pid))
+                    .format(root_process.cmdline(), root_process.pid))
+        _, running_descendants = kill_processes(logger, running_descendants,
+                                                sig=signal.SIGTERM, timeout=timeout)
     else:
-        logger.debug("There are no descendant processes to kill")
+        logger.debug("There are no descendant processes to terminate.")
+
+    logger.info("Terminating root process PID: {}.".format(root_process.pid))
```

Comment:
@jgao54 I don't understand this part. The `root_process` is the monitoring `--local` task which is the `parent` of the `--raw` task and its underlying processes. So it is trying to SIGTERM itself here and SIGKILL eventually. That doesn't make sense.

If `--local` finds the need to terminate its children, it should exit itself gracefully maybe with a non 0 exitcode.